pkg/cluster/clustermsi.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].ClientID == nil {
+		return errors.New("found nil ClientID while parsing cluster MSI CredentialsObject")
+	}
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].TenantID == nil {
+		return errors.New("found nil TenantID while parsing cluster MSI CredentialsObject")
+	}
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].ClientSecret == nil {
+		return errors.New("found nil ClientSecret while parsing cluster MSI CredentialsObject")
+	}
+
+	clientId := *kvSecret.CredentialsObject.ExplicitIdentities[0].ClientID
+	tenantId := *kvSecret.CredentialsObject.ExplicitIdentities[0].TenantID
+	certData := *kvSecret.CredentialsObject.ExplicitIdentities[0].ClientSecret
+
+	decodedCertData, err := base64.StdEncoding.DecodeString(certData)
+	if err != nil {
+		return err
+	}
+
+	certs, key, err := azidentity.ParseCertificates(decodedCertData, nil)
+	if err != nil {
+		return err
+	}
+
+	cred, err := azidentity.NewClientCertificateCredential(tenantId, clientId, certs, key, m.env.Environment().ClientCertificateCredentialOptions())
+	if err != nil {
+		return err
+	}
+
+	// Note that we are assuming that all of the platform MIs are in the same subscription.
+	resourceId, err := arm.ParseResourceID(m.doc.OpenShiftCluster.Properties.PlatformWorkloadIdentityProfile.PlatformWorkloadIdentities[0].ResourceID)
+	if err != nil {
+		return err
+	}
+
+	subId := resourceId.SubscriptionID
+	clientOptions := getClientOptions(m.env.Environment())
+	clusterMsiFederatedIdentityCredentials, err := armmsi.NewFederatedIdentityCredentialsClient(subId, cred, &clientOptions)
+	if err != nil {
+		return err
+	}
+
+	m.clusterMsiFederatedIdentityCredentials = clusterMsiFederatedIdentityCredentials
+	return nil
+}
+
+// clusterMsiSecretName returns the name to store the cluster MSI certificate under in
+// the cluster MSI key vault.
+func (m *manager) clusterMsiSecretName() (string, error) {
+	clusterMsi, err := m.clusterMsiResourceId()
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s", m.doc.ID, clusterMsi.Name), nil
+}
+
+// clusterMsiResourceId returns the resource ID of the cluster MSI or an error
+// if it encounters an issue while grabbing the resource ID from the cluster
+// doc. It is written under the assumption that there is only one cluster MSI
+// and will have to be refactored if we ever use more than one.
+func (m *manager) clusterMsiResourceId() (*arm.ResourceID, error) {
+	var clusterMsi *arm.ResourceID
+	if m.doc.OpenShiftCluster.Identity != nil && m.doc.OpenShiftCluster.Identity.UserAssignedIdentities != nil {
+		for msiResourceId := range m.doc.OpenShiftCluster.Identity.UserAssignedIdentities {
+			var err error
+			clusterMsi, err = arm.ParseResourceID(msiResourceId)
+			if err != nil {
+				return nil, err
+			}
+		}
```

Comment:
So, we always pick the last non errored out MSI. Can we pick the first non-errored out MSI instead to save some cycles? Lmk, if this would change the actual behavior for any reason.